docs/src/references/agent/operation-workflow.md

```diff
@@ -264,26 +246,264 @@ The script exit status and output is used to determine the next step for the com
+
+```toml
+["<state>"]
+action = "<action-name> <possibly-with-arguments>"
+```
+
+As for scripts, these actions can be complemented with handlers dictating how to proceed when the action is successful or failed.
+The set of accepted handlers for an action are the following:
+
+- `on_success = "<state>"` defines the next state when the action is successful
+- `on_error = { status = "<state>", reason = "<message>"}` defines the next state when the action failed
+- `timeout_second = 3600` the number of second given to the action to execute
+- `on_timeout = { status = "<state>", reason = "<message>"}` defines the next state when the action is not be completed within the time limit 
+
+For some action, notably a device `restart`, the handlers are limited to one:
+- `on_exec = "<state>"` defines the next state once the action has been launched in the background.
+  The action outcome will have to be observed in this `on_exec` state.
+
+Currently, here are the available actions:
+
+- `restart` triggers a reboot of the device
+- `waiting <delegate>` means that agent itself has nothing to do but to wait that another component,
+  to which the action has been delegated, finalizes its task and notifies the new state for the command.  
+- `builtin` is used when a builtin operation is overwritten by a custom workflow and indicates that for that state
+  the builtin action has to be applied.
+- `proceed` is a no-op action, simply proceeding to the next state, which is useful when a builtin operation is customized
+   but no specific behavior has to be added on a workflow extension point.
+- `cleanup` marks the terminal state of the workflow where the command has been fully processed
+  and where the original requester is expected to clean up the command retained message storing its state.
+
+#### Restart
 
 A workflow can trigger a device restart, using the builtin __restart__ action.
 
 This action is controlled by three states:
-- the *on_executing* state to which the workflow moves before the reboot is triggered
+- the *on_exec* state to which the workflow moves before the reboot is triggered
 - the *on_success* state to which the workflow resumes after a successful device reboot
 - the *on_error* state to which the workflow resumes in case the reboot fails
 
 For instance, the following triggers a reboot from the `reboot_required` state
 and moves to `restarting` waiting for the device to restart
 and finally to either `successful_restart` or `failed_restart`,
-depending on the actual status of the reboot. 
+depending on the actual status of the reboot.
+
+```toml
+["device-restart"]
+action = "restart"
+on_exec = "waiting-for-restart"
+on_success = "successful_restart"
+on_error = "failed_restart"
+```
+
+Alternative proposal:
+
+```toml
+["device-restart"]
+action = "restart"
+on_exec = "waiting-for-restart"
+
+["waiting-for-restart"]
+action = "waiting restart"
+on_success = "successful_restart"
+on_error = "failed_restart"
+```
```

Comment:
Unfortunately, the second is not implemented yet. The main reason being that the `"waiting restart"` needs more thinking. The failing system test (aka "Trigger native-reboot within workflow (on_error) - missing sudoers entry for reboot"), highlights one of these points: the behavior in case of an error is not clear.